src/main/java/backend/Logic.java

```diff
@@ -118,17 +135,26 @@ public void refresh() {
 
             logger.info("Opening " + repoId);
             UI.status.displayMessage("Opening " + repoId);
-            Platform.runLater(() -> UI.events.triggerEvent(new RepoOpeningEvent(repoId, isPrimaryRepository)));
+
+            Platform.runLater(() -> UI.events.triggerEvent(new FilterRepoOpeningEvent()));
+            if (isPrimaryRepository) {
+                Platform.runLater(() -> UI.events.triggerEvent(new PrimaryRepoOpeningEvent()));
+            }
 
             return repoOpControl.openRepository(repoId)
                     .thenApply(models::addPending)
                     .thenRun(this::refreshUI)
-                    .thenRun(() ->
+                    .thenRun(() -> {
+                        Platform.runLater(() -> UI.events.triggerEvent(new FilterRepoOpenedEvent()));
+                        if (isPrimaryRepository) {
                             Platform.runLater(() ->
-                                    // to trigger the event from the UI thread
-                                    UI.events.triggerEvent(new RepoOpenedEvent(repoId, isPrimaryRepository))
-                            )
-                    )
+                                    UI.events.triggerEvent(new PrimaryRepoOpenedEvent()));
+                            return;
+                        }
+
+                        Platform.runLater(() ->
+                                UI.events.triggerEvent(new AppliedFilterEvent(panel.get())));
+                    })
```

Comment:
This part (the `lambda`) can be extracted into a method too. It's a bit nested here and doesn't seem to flow well with the rest of the `Future`s in the chain.